Update docker

gonuke opened the pull request:
## Description
Updates the Dockerfile to build with more recent versions of dependencies.  Specifically,
* all references to `long` are removed.  They became identical to `int` in a recent version and now create an error, specifically in Cython compilations
* add a `make_cores` Docker build_arg to support variable sized parallel builds at all stages
* updated syntax of Docker `ENV` command
* newer standard version of HDF5 (with newer well-formed version string)
* add a `dagmc_version` Docker build_arg and set a new version as default
* force OpenMC to install in `--user` mode

## Motivation and Context
Build under CI had become unreliable.

## Changes
This addresses bugs in the Docker build

gonuke committed:
remove reference to python `long`

gonuke committed:
Improvements to standard Dockerfile, including:

* update ENV syntax
* add `make_cores` build arg
* unpin numpy version
* update HDF5 version and version string
* introduce DAGMC version build arg
* update OpenMC version
* force --user install of OpenMC

gonuke committed:
spread make_cores around and back off OpenMC version

gonuke committed:
update CHANGELOG

gonuke committed:
add PR number to CHANGELOG

gonuke committed:
update PR #

ahnaf-tahmid-chowdhury commented on docker/ubuntu_22.04-dev.dockerfile:
```diff
@@ -108,31 +114,33 @@ RUN cd /root \
     git clone --depth 1 --branch $openmc_version https://github.com/openmc-dev/openmc.git $HOME/opt/openmc \
     && cd  $HOME/opt/openmc \
-    && pip install .
+    && pip install --user .
```
Perhaps we can use a virtual environment.

ahnaf-tahmid-chowdhury commented on docker/ubuntu_22.04-dev.dockerfile:
```diff
@@ -108,31 +114,33 @@ RUN cd /root \
 COPY . $HOME/opt/pyne
 RUN export PYNE_HDF5_ARGS="--hdf5 $HDF5_INSTALL_PATH"; \
     cd $HOME/opt/pyne \
     && python setup.py install --user \
```
This user prefix is not the same as the standard one set by pip. PyNE uses custom logic for it, which is causing a conflict.

ahnaf-tahmid-chowdhury reviewed (changes requested):
Thanks for the PR. I think using a virtual env may fix the issue, at least for now.

ahnaf-tahmid-chowdhury commented:
The current actions are failing, saying that python is not found.

gonuke commented:
> The current actions are failing, saying that python is not found.

I think that's for conda - I'm investigating now

ahnaf-tahmid-chowdhury commented:
Yes, conda is not activated. It's trying to use the system python. We may need to activate the base environment or use `conda run python setup.py`.

gonuke commented:
> Yes, conda is not activated. It's trying to use the system python. We may need to activate the base environment or use `conda run python setup.py`.

That's not the problem.  You don't need to activate a conda environment if you have never established a new environment.  For some reason, it seems that installing openmc is causing python to be uninstalled from conda!!

gonuke committed:
cleanup warning

gonuke committed:
Turn off spatial solvers by default.

The AHOT spatial solvers are poorly supported and tangential
to the primiary value of PyNE.  In addition, recent compiler
compatibility can cause problems during compilation.

While it may make sense to remove them entirely,
they will be turned off by default for now.

ahnaf-tahmid-chowdhury committed:
freeze-install python

ahnaf-tahmid-chowdhury committed:
use CONDA_PREFIX

gonuke committed:
update Docker ENV syntax

ahnaf-tahmid-chowdhury commented:
We're encountering a `Fatal Python error: Illegal instruction` in the `test_dagmc.py`. This usually points to a hardware-specific instruction being used that CPU doesn't support or an incompatible binary being run.

ahnaf-tahmid-chowdhury committed:
prefix /opt/conda

ahnaf-tahmid-chowdhury committed:
use pytest -ra

gonuke commented:
> We're encountering a `Fatal Python error: Illegal instruction` in the `test_dagmc.py`. This usually points to a hardware-specific instruction being used that CPU doesn't support or an incompatible binary being run.

I'm pretty sure this is because OpenMC is being installed with DAGMC compiled with DoubleDown and Embree, which includes processor-specific instructions that may not be available on all platforms.  One solution is to build/tests without dagmc
```
"conda-forge::openmc=*=*nodagmc*"
```

I tested this and it works for me locally

ahnaf-tahmid-chowdhury committed:
better format

gonuke commented on docker/conda_build.dockerfile:
```diff
@@ -1,92 +1,74 @@
+ENV TZ=America/Chicago \
+    HOME=/root \
+    PATH=/opt/conda/bin:$PATH \
+    LD_LIBRARY_PATH=/opt/conda/lib:$LD_LIBRARY_PATH
```
I think docker now complains (warning) that `LD_LIBRARY_PATH` is not yet defined, so it's probably better to do this:
```suggestion
    LD_LIBRARY_PATH=/opt/conda/lib
```

gonuke reviewed (commented):
Thanks @ahnaf-tahmid-chowdhury - this looks like it might be ready now

ahnaf-tahmid-chowdhury committed:
Update docker/conda_build.dockerfile

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

ahnaf-tahmid-chowdhury reviewed (approved):
Let me know if there are further changes needed.
